cmd/libsnap-confine-private/snap.c

```diff
@@ -97,6 +97,112 @@ static int skip_one_char(const char **p, char c)
 	return 0;
 }
 
+void sc_instance_name_validate(const char *instance_name,
+			       struct sc_error **errorp)
+{
+	// NOTE: This function should be synchronized with the two other
+	// implementations: validate_instance_name and snap.ValidateInstanceName.
+	struct sc_error *err = NULL;
+
+	// Ensure that name is not NULL
+	if (instance_name == NULL) {
+		err =
+		    sc_error_init(SC_SNAP_DOMAIN, SC_SNAP_INVALID_INSTANCE_NAME,
+				  "snap instance name cannot be NULL");
+		goto out;
+	}
+
+	const char *instance_sep = strchr(instance_name, '_');
+
+	if (instance_sep != NULL) {
+		size_t snap_name_len = instance_sep - instance_name;
+		const char *instance_key = instance_sep + 1;
+
+		if (snap_name_len == 0) {
+			err =
+			    sc_error_init(SC_SNAP_DOMAIN,
+					  SC_SNAP_INVALID_INSTANCE_NAME,
+					  "snap name must contain at least one letter");
+			goto out;
+		}
+
+		sc_instance_key_validate(instance_key, &err);
+		if (err != NULL) {
+			goto out;
+		}
+
+		if (snap_name_len > 40) {
+			err =
+			    sc_error_init(SC_SNAP_DOMAIN,
+					  SC_SNAP_INVALID_INSTANCE_NAME,
+					  "snap name must be shorter than 40 characters");
+			goto out;
+		}
+
+		char *snap_name SC_CLEANUP(sc_cleanup_string) =
+		    strndup(instance_name, snap_name_len);
+		sc_snap_name_validate(snap_name, &err);
+		if (err != NULL) {
+			goto out;
+		}
+
+	} else {
+		// just snap name
+		sc_snap_name_validate(instance_name, &err);
+	}
+
+ out:
+	sc_error_forward(errorp, err);
+}
+
+void sc_instance_key_validate(const char *instance_key,
+			      struct sc_error **errorp)
+{
+	// NOTE: see snap.ValidateInstanceName for reference of a valid instance key
+	// format
+	struct sc_error *err = NULL;
+
+	// Ensure that name is not NULL
+	if (instance_key == NULL) {
+		err = sc_error_init(SC_SNAP_DOMAIN, SC_SNAP_INVALID_NAME,
+				    "instance key cannot be NULL");
+		goto out;
+	}
+
+	const char *p = instance_key;
```

Comment:
See my comment on this in https://github.com/snapcore/snapd/pull/5584#discussion_r206544629 - same as above, I wonder if we can share this code. Having two C implementations is a bit annoying.

Reply:
Yeah, you redid the implementation in PR 5584 with some nice cleanups, it would be great to have the equivalent changes here.